GradleUtils.getDependencyProject not working for typesafe project accessors

When Minecraft is being set up loom tries to resolve the mods that are added as dependencies into the project. However, when adding a dependency using a typesafe project accessor like:
```kts
"runtimeImpl"(projects.ext.adventureHelper)
```
the code breaks while using reflection to get the dependency project. This seems to be the case as typesafe project accessors actually don't extend `DefaultProjectDependency` directly but rather extend `DelegatingProjectDependency`. This leads to the following error being raised:
```
Caused by: java.lang.IllegalArgumentException: Can not get final org.gradle.api.internal.project.ProjectInternal field org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency.dependencyProject on org.gradle.accessors.dm.Ext_AdventureHelperProjectDependency
	at java.base/jdk.internal.reflect.MethodHandleFieldAccessorImpl.newGetIllegalArgumentException(MethodHandleFieldAccessorImpl.java:86)
	at java.base/jdk.internal.reflect.MethodHandleObjectFieldAccessorImpl.get(MethodHandleObjectFieldAccessorImpl.java:61)
	at net.fabricmc.loom.util.gradle.GradleUtils.getDependencyProject(GradleUtils.java:102)
	at net.fabricmc.loom.configuration.processors.SpecContextImpl.getDependentMods(SpecContextImpl.java:89)
	at net.fabricmc.loom.configuration.processors.SpecContextImpl.create(SpecContextImpl.java:63)
	at net.fabricmc.loom.configuration.processors.MinecraftJarProcessorManager.create(MinecraftJarProcessorManager.java:70)
	at net.fabricmc.loom.configuration.CompileConfiguration.setupMinecraft(CompileConfiguration.java:177)
	at net.fabricmc.loom.configuration.CompileConfiguration.lambda$run$1(CompileConfiguration.java:109)
	... 174 more
``` 

It seems like the project accessors are generated [here](https://github.com/gradle/gradle/blob/53afef85c1c36406b0e5132d6be7f306ff451fb4/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/catalog/ProjectAccessorsSourceGenerator.java) which shows that those are extending DelegatingProjectDependency and are not a DefaultProjectDependency directly.

I wasn't sure if just catching the exception or trying to resolve the dependency by recursively trying to resolve via the `delegate` field was the better solution, so I thought I'm just gonna open the issue instead. If you want, I would take a look into proving a pull request to resolve the issue as well.

--- a/src/main/java/net/fabricmc/loom/util/gradle/GradleUtils.java
+++ b/src/main/java/net/fabricmc/loom/util/gradle/GradleUtils.java
@@ -32,10 +32,15 @@ import org.gradle.api.Project;
 import org.gradle.api.artifacts.ProjectDependency;
 import org.gradle.api.file.RegularFileProperty;
 import org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency;
+import org.gradle.api.internal.catalog.DelegatingProjectDependency;
 import org.gradle.api.invocation.Gradle;
 import org.gradle.api.provider.Provider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public final class GradleUtils {
+	private static final Logger LOGGER = LoggerFactory.getLogger(GradleUtils.class);
+
 	private GradleUtils() {
 	}
 
@@ -95,14 +100,29 @@ public final class GradleUtils {
 	// Get the project from the field with reflection to suppress the deprecation warning.
 	// If you hate it find a solution yourself and make a PR, I'm getting a bit tired of chasing Gradle updates
 	public static Project getDependencyProject(ProjectDependency projectDependency) {
-		try {
-			final Class<DefaultProjectDependency> clazz = DefaultProjectDependency.class;
-			final Field dependencyProject = clazz.getDeclaredField("dependencyProject");
-			dependencyProject.setAccessible(true);
-			return (Project) dependencyProject.get(projectDependency);
-		} catch (NoSuchFieldException | IllegalAccessException ignored) {
-			// Just fallback and trigger the warning, this will break in Gradle 9
-			return projectDependency.getDependencyProject();
+		if (projectDependency instanceof DefaultProjectDependency) {
+			try {
+				final Class<DefaultProjectDependency> clazz = DefaultProjectDependency.class;
+				final Field dependencyProject = clazz.getDeclaredField("dependencyProject");
+				dependencyProject.setAccessible(true);
+				return (Project) dependencyProject.get(projectDependency);
+			} catch (NoSuchFieldException | IllegalAccessException e) {
+				LOGGER.warn("Failed to reflect DefaultProjectDependency", e);
+			}
+		} else if (projectDependency instanceof DelegatingProjectDependency) {
+			try {
+				final Class<DelegatingProjectDependency> clazz = DelegatingProjectDependency.class;
+				final Field delgeate = clazz.getDeclaredField("delegate");
+				delgeate.setAccessible(true);
+				return getDependencyProject((ProjectDependency) delgeate.get(projectDependency));
+			} catch (NoSuchFieldException | IllegalAccessException e) {
+				LOGGER.warn("Failed to reflect DelegatingProjectDependency", e);
+			}
 		}
+
+		// Just fallback and trigger the warning, this will break in Gradle 9
+		final Project project = projectDependency.getDependencyProject();
+		LOGGER.warn("Loom was unable to suppress the deprecation warning for ProjectDependency#getDependencyProject, if you are on the latest version of Loom please report this issue to the Loom developers and provide the error above, this WILL stop working in a future Gradle version.");
+		return project;
 	}
 }

--- a/src/test/groovy/net/fabricmc/loom/test/unit/GradleUtilsTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/unit/GradleUtilsTest.groovy
@@ -1,0 +1,59 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2024 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.test.unit
+
+import org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency
+import org.gradle.api.internal.catalog.DelegatingProjectDependency
+import org.gradle.api.internal.project.ProjectInternal
+import spock.lang.Specification
+
+import net.fabricmc.loom.util.gradle.GradleUtils
+
+class GradleUtilsTest extends Specification {
+	def "get default project dependency"() {
+		given:
+		def project = Mock(ProjectInternal)
+		def dependency = new DefaultProjectDependency(project, false)
+
+		when:
+		def result = GradleUtils.getDependencyProject(dependency)
+
+		then:
+		result == project
+	}
+
+	def "get delegated project dependency"() {
+		given:
+		def project = Mock(ProjectInternal)
+		def dependency = new DefaultProjectDependency(project, true)
+		def delegate = new DelegatingProjectDependency(null, dependency)
+
+		when:
+		def result = GradleUtils.getDependencyProject(delegate)
+
+		then:
+		result == project
+	}
+}

